core/chains/evm/logpoller/log_poller.go

```diff
@@ -881,22 +881,53 @@ func (lp *logPoller) PollAndSaveLogs(ctx context.Context, currentBlockNumber int
 	}
 }
 
+// Returns information about latestBlock, latestFinalizedBlockNumber
+// If finality tag is not enabled, latestFinalizedBlockNumber is calculated as latestBlockNumber - lp.finalityDepth (configured param)
+// Otherwise, we return last finalized block number returned from chain
+func (lp *logPoller) latestBlocks(ctx context.Context) (*evmtypes.Head, int64, error) {
+	// If finality is not enabled, we can only fetch the latest block
+	if !lp.useFinalityTag {
+		// Example:
+		// finalityDepth = 2
+		// Blocks: 1->2->3->4->5(latestBlock)
+		// lastFinalizedBlockNumber would be 3
+		latestBlock, err := lp.ec.HeadByNumber(ctx, nil)
+		if err != nil {
+			return nil, 0, err
+		}
+		// If chain has fewer blocks than finalityDepth, return 0
+		return latestBlock, mathutil.Max(latestBlock.Number-lp.finalityDepth, 0), nil
+	}
+
+	// If finality is enabled, we need to get the latest and finalized blocks.
+	blocks, err := lp.batchFetchBlocks(ctx, []string{rpc.LatestBlockNumber.String(), rpc.FinalizedBlockNumber.String()}, 2)
+	if err != nil {
+		return nil, 0, err
+	}
+	latest := blocks[0]
+	finalized := blocks[1]
+	return latest, finalized.Number, nil
+}
+
 // Find the first place where our chain and their chain have the same block,
 // that block number is the LCA. Return the block after that, where we want to resume polling.
-func (lp *logPoller) findBlockAfterLCA(ctx context.Context, current *evmtypes.Head) (*evmtypes.Head, error) {
+func (lp *logPoller) findBlockAfterLCA(ctx context.Context, current *evmtypes.Head, lastFinalizedBlockNumber int64) (*evmtypes.Head, error) {
 	// Current is where the mismatch starts.
 	// Check its parent to see if its the same as ours saved.
 	parent, err := lp.ec.HeadByHash(ctx, current.ParentHash)
 	if err != nil {
 		return nil, err
 	}
+	// If finalityTag is not used, we need to compute lastFinalizedBlockNumber based on the configured finalityDepth and the parent block
+	if !lp.useFinalityTag {
+		lastFinalizedBlockNumber = parent.Number - lp.finalityDepth
```

Comment:
I think this should be:
`lastFinalizedBlockNumber = current.Number - lp.finalityDepth`

and then below we want:

`for parent.Number >= lastFinalizedBlockNumber - 1 {`

Reply:
At least, that's what would keep present logic the same... although it doesn't seem consistent with this comment:
```
	// If the parent block number becomes < the first finalized block our reorg is too deep.
```
which would indicate to me either the comment was wrong or the current logic is 1 block safer than we need?

Reply:
Hmm, I think that
```go
lastFinalizedBlockNumber = parent.Number - lp.finalityDepth
```
is the same as 
```go
lastFinalizedBlockNumber = current.Number - lp.finalityDepth - 1
// because `current` is one ahead of the parent
```


I suggest to do the following.

```go
lastFinalizedBlockNumber = current.Number - lp.finalityDepth - 1

// Keep the for loop as it is
for parent.Number >= lastFinalizedBlockNumber  { }
```

Do you think it makes sense?

Reply:
IIUC, what we are trying to do here is to find the first block (after the last finalized block) that matches our state in the database. Worst case scenario, we should return the last finalized block, so the loop should terminate when `parent.Number < lastFinalizedBlockNumber`, which matches current implementation 

> I  think this should be:
`lastFinalizedBlockNumber = current.Number - lp.finalityDepth`
and then below we want:
`for parent.Number >= lastFinalizedBlockNumber - 1 {`

I think that `parent.Number - lp.finalityDepth` is actually `current.Number - lp.finalityDepth - 1` (because parent is one block behind), so it kind of matches current logic, no?

Reply:
Although, I'm not sure why did we compute max reorg depth based on the `parent` block, not the `current` in the previous implementation
```go
for parent.Number >= (reorgStart - lp.finalityDepth)  // reorgStart = parent.Number
```
Was that accidental?

Reply:
> Was that accidental

yeah looks to me like its one block extra of safety for no reason

Reply:
Removed extra block, it required increasing finality depth in tests

Reply:
Alright, makes sense now!